### PR TITLE
feat: sign in with token endpoint

### PR DESCRIPTION
**Pull request by Timmy-id**
## What does the PR do?

This PR is about requesting and using an OTP to log in. The endpoint returns errors if:

- [x] User is not registered.
- [x] User inputs the wrong OTP.
- [x] Validation errors on the user inputs invalid or wrong details

How should this be manually tested?

1. Enter your email to request for OTP

2. Make a POST request to the `\api\v1\auth\signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
}
```

3. A 6-digit OTP is then sent to the user's registered email

4. Then, make a POST request to the `\api\v1\auth\verify-signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
  "token": "123456"
}
```

Successful Response

On a successful testimonial creation, the API should return a 200 OK status code.

```json
    {
      message: 'Sign-in token sent to email',
      status_code: 200,
    }
```

```json
    {
      message: 'Sign-in successful',
      "token": "eyJhbGciOiJI.......",
      "status_code": 200
    }
```

Authentication Error:

When the user is not registered:

```json
{
  "message": "User not found",
  "status_code": 404
}
```

Invalid OTP error

On Invalid OTP error, if the user is putting an invalid OTP or email:

```json
{
  "message": "Invalid token or email",
  "status_code": 401
}
```

Validation:

The response should return a 400 Bad Request status code with appropriate validation error messages if any required field is missing or invalid.

```json
{
  "message": ["email must be a string", "email must be an email"],
  "status_code": "400"
}
```

Task

- [x] Create a POST endpoint POST /api/v1/auth/signin-token/ to request for OTP for registered users.
- [x] Implement server-side logic to handle successful and failed update requests.
- [x] Ensure the registered user receives an OTP in their email.
- [x] The OTP is properly saved in the database by hashing it
- [x] Check that the OTP is valid
- [x] Then the user gets a token after all validations are passed

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/140)

### CODE REVIEW

**Comment by Timmy-id** [comment]
@Homoakin619 I've fixed my code.

**Comment by Timmy-id** [comment]
@Heba-WebDev Help review

**Comment by CFCIfe** [comment]
> On a successful testimonial creation, the API should return a 200 OK status code.

On a successful token creation, the API should return a 200 OK status code.

File: package.json [hunk 1]
```diff
@@ -66,7 +65,7 @@
     "@nestjs/cli": "^10.0.0",
     "@nestjs/schematics": "^10.0.0",
     "@nestjs/testing": "^10.0.0",
-    "@types/bcrypt": "^5.0.2",
+    "@types/bcryptjs": "^2.4.6",
```
Comment (CFCIfe) [inline package.json:68]: Any reason we are changing this?

File: src/modules/auth/auth.controller.ts [hunk 2]
```diff
@@ -58,4 +60,22 @@ export default class RegistrationController {
     const { id: user_id } = request['user'];
     return this.authService.enable2FA(user_id, password);
   }
+
+  @skipAuth()
+  @Post('signin-token')
+  @ApiOperation({ summary: 'Request Signin Token' })
+  @ApiResponse({ status: 200, description: 'Sign-in token sent to email', type: RequestSigninTokenDto })
+  @ApiResponse({ status: 400, description: 'Bad request' })
+  public async signInToken(@Body() body: RequestSigninTokenDto, @Res() response: Response) {
```
Comment (CFCIfe) [inline src/modules/auth/auth.controller.ts:69]: You don't seem to need the response decorator here

```suggestion
public async signInToken(@Body() body: RequestSigninTokenDto) {
```

File: src/modules/auth/auth.controller.ts [hunk 3]
```diff
@@ -58,4 +60,22 @@ export default class RegistrationController {
     const { id: user_id } = request['user'];
     return this.authService.enable2FA(user_id, password);
   }
+
+  @skipAuth()
+  @Post('signin-token')
+  @ApiOperation({ summary: 'Request Signin Token' })
+  @ApiResponse({ status: 200, description: 'Sign-in token sent to email', type: RequestSigninTokenDto })
+  @ApiResponse({ status: 400, description: 'Bad request' })
```
Comment (CFCIfe) [inline src/modules/auth/auth.controller.ts:68]: Describe Validation error

```suggestion
@ApiResponse({ status: 400, description: 'Bad request' })
@ApiResponse({ status: 401, description: 'Unauthorized request' })
```

File: src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -257,4 +268,66 @@ export default class AuthenticationService {
...
+        message: USER_NOT_FOUND,
+        status_code: HttpStatus.NOT_FOUND,
+      });
+    }
+
+    const otpExist = await this.otpService.findOtp(user.id);
+
+    const newOtp = generateSixDigitToken();
+
+    if (otpExist) {
```
Comment (CFCIfe) [inline src/modules/auth/auth.service.ts:288]: If there is a situation that otpExist will be false, we should throw a descriptive error.

I checked the findOtp method, and it does not handle this as well.

File: src/modules/auth/auth.controller.ts (hunk 2 above)
Comment (Timmy-id) [reply src/modules/auth/auth.controller.ts:69]: Fixed

File: src/modules/auth/auth.service.ts [hunk 5]
```diff
@@ -257,4 +268,66 @@ export default class AuthenticationService {
...
+
+      await this.otpService.createOtp(user.id);
+
+      await this.emailService.sendLoginOtp(user.email, newOtp);
+
+      return;
+    }
+    await this.otpService.createOtp(user.id);
+
+    await this.emailService.sendLoginOtp(user.email, newOtp);
```
Comment (CFCIfe) [inline src/modules/auth/auth.service.ts:299]:
```suggestion
if (otpExist) {
    await this.otpService.deleteOtp(user.id);
  }
  
  // Generate a new OTP and save it
  const newOtp = generateSixDigitToken();
  await this.otpService.createOtp(user.id);

  // Send the OTP to the user's email
  await this.emailService.sendLoginOtp(user.email, newOtp);
```

Comment (CFCIfe) [inline src/modules/auth/auth.service.ts:299]:
```suggestion
if (otpExist) {
    await this.otpService.deleteOtp(user.id);
  }
  
  // Generate a new OTP and save it
  const newOtp = generateSixDigitToken();
  await this.otpService.createOtp(user.id);

  // Send the OTP to the user's email
  await this.emailService.sendLoginOtp(user.email, newOtp);
```

Comment (Timmy-id) [reply src/modules/auth/auth.service.ts:299]: Fixed

Comment (Timmy-id) [reply src/modules/auth/auth.service.ts]: @CFCIfe I have made the required changes. Help review.

File: package.json (hunk 1 above)
Comment (Timmy-id) [reply package.json]: Fixed @CFCIfe

File: src/modules/auth/auth.service.ts
Comment (Heba-WebDev) [inline src/modules/auth/auth.service.ts:1]: When the user is not found, no need to tell the client (can be an attacker) about that. Just throw a bad request for invalid credentials.

Comment (Timmy-id) [reply src/modules/auth/auth.service.ts:1]: Fixed @Heba-WebDev

**Review by Heba-WebDev — DISMISSED** [review]

File: src/modules/auth/auth.service.ts
Comment (Heba-WebDev) [reply src/modules/auth/auth.service.ts:1]: Return 200 not bad request

**Review by davidolanrewaju — DISMISSED** [review]

File: src/modules/auth/auth.controller.ts [hunk 6]
```diff
@@ -58,4 +60,24 @@ export default class RegistrationController {
...
+  @Post('signin-token')
+  @HttpCode(200)
+  @ApiOperation({ summary: 'Request Signin Token' })
+  @ApiResponse({ status: 200, description: 'Sign-in token sent to email', type: RequestSigninTokenDto })
+  @ApiResponse({ status: 400, description: 'Bad request' })
+  public async signInToken(@Body() body: RequestSigninTokenDto) {
+    return await this.authService.requestSignInToken(body);
+  }
+
+  @skipAuth()
```
Comment (Homoakin619) [inline src/modules/auth/auth.controller.ts:74]: I believe this routes are protected if yes, remove the skipAuth decorator, the decorator is for bypassing authorization. By default all routes are protected

File: src/modules/auth/auth.controller.ts [hunk 7]
```diff
@@ -58,4 +60,24 @@ export default class RegistrationController {
     const { id: user_id } = request['user'];
     return this.authService.enable2FA(user_id, password);
   }
+
+  @skipAuth()
```
Comment (Homoakin619) [inline src/modules/auth/auth.controller.ts:64]: If this route is protected, remove this decorator

File: src/modules/auth/auth.controller.ts (hunk 6 above)
Comment (Timmy-id) [reply src/modules/auth/auth.controller.ts:74]: fixed

File: src/modules/auth/auth.controller.ts (hunk 7 above)
Comment (Timmy-id) [reply src/modules/auth/auth.controller.ts:64]: This route is not protected

**Review by Cobidex — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by TheCodeGhinux — APPROVED** [review]

**Pull request merged**